### ISSUE TITLE
Playmaker and Fungus Integration (Back-and-forth actions/blocks)

### ISSUE THREAD

**Issue by astraldata** [snozbot/fungus#547]
Seems like it wouldn't be hard to add, but actions to support back-and-forth between Playmaker FSMs would be really fun to use with Fungus, and the simplicity of Playmaker would lend itself really well to story-based gameplay.

Some custom Playmaker actions and custom Fungus action-blocks to back-and-forth between the two would be excellent!

**Comment by chrisgregan** [comment]
Prob won't have time myself, but if anyone is looking to take this one on, take a look at Fungus/Integrations/Spine for an example of how we support this type of integration. The idea is to bundle a .unitypackage to avoid users who don't have PlayMaker installed getting error messages.

**Comment by chrisgregan** [comment]
There are a few obvious things that would be beneficial here:
1. A PlayMaker action which can execute a named Fungus block in a Flowchart.
2. A Fungus command which can trigger a PlayMaker state transition in a FSM.
3. A custom Fungus command to copy a Fungus variable value to a PlayMaker variable.
4. A custom PlayMaker action to copy a PlayMaker variable to a Fungus command.

I haven't looked at PlayMaker in a while, but there might be other types of possible interaction between the two systems that would be benificial, so if you have ideas post em here.

As noted in a previous comment, for the final deployment the integration code should be contained in a .unitypackage so that clients can choose to install it if they are using PlayMaker (otherwise the integration code would cause compile errors if PlayMaker is not present.)

**Comment by Nilihum** [comment]
Been looking over this for the past week because I've decided to try and accomplish this as part of my fourth year project. I'm going to be on a deadline for it though, so I was wondering if there is a good starting point to tackle this or a minimum amount of functionality I can implement and then build upon, adding more and more as I go?

As for the four beneficial things Chris has brought up; how do these tie into each other or are they four mostly independent things?

**Comment by Nilihum** [comment]
I've been looking further into this project by looking over the code in Fungus/Integration/Spine and I'm kind of at a loss on how to approach this, not helped in the least that the Dr Jeckyl example scene has missing prefabs and the associated files have errors, (possibly due to using Spine.Unity;).

Can anyone give me any advice such as a starting point or some sort of guide? Is this project even possible to accomplish in the space of three months? Thanks in advance.

**Comment by chrisgregan** [comment]
@Nilihum The readme.txt file in the Fungus/Integration/Spine folder explains that you first need to install the Spine runtime for Unity. If you skip that step you're going to get compile errors. We're not permitted to distribute the Spine runtime ourselves, hence the need for the Unity Package.

For a starting point, install Fungus and PlayMaker in the same project. Learn how to create a [custom action](https://hutonggames.fogbugz.com/?W166) in PlayMaker. Learn how to create a [custom command](https://www.youtube.com/watch?v=WFEGnS7MU74) in Fungus. Learn also how to [execute a Block](http://fungusdocs.snozbot.com/coding_tips.html#executing_blocks) via the flowchart class.

From there, you should be able to use a custom PlayMaker action to execute a Block in Fungus, and vice versa. Once you get that working, move onto getting variables from the two systems to interact. I pointed you to the Spine integration mainly as an example of how you should package up the finished components for people to use in their own projects, which is the last step in the process.

The time scale depends on your level of experience with Unity. If you work your way through the tutorials carefully and do a bit of experimentation on your own to find out how the systems work then it should be doable with a few days works. If you're fairly new to Unity then I would recommend working through the Unity [scripting tutorials](https://unity3d.com/learn/tutorials/topics/scripting) first to see how things work.

**Comment by Nilihum** [comment]
I've only got the chance to work on this recently, hence why I've not been updating, but I just thought I'd drop a line and give an update. In the passed few days I have managed to create a Playmaker action that can write a string into a string in Fungus, and intend to expand this to accept other variable types. I've also created a Playmaker action that can execute named blocks in a given flowchart.

The trouble I'm having now is accessing  states within FSMs with is preventing me from accessing variables or causing state transitions. Th [custom actions guide](https://hutonggames.fogbugz.com/?W166) on Hutong Games is a bit difficult to parse and I was wondering if anyone had any better tutorials I could use or if they had any experience with Playmaker that they could share.

**Comment by Nilihum** [comment]
Right, so, assignments and deadlines have been keeping me away, but I've got most of this done. I've created a playmaker action that can execute Fungus blocks and I can write strings, integers, floats, and booleans to Fungus. I've also managed to send global transitions to the Playmaker FSM using Fungus, and while I don't have a script that can write in variables from Fungus to Playmaker, I have a Playmaker script that reads from Fungus variables and places the values into Playmaker.

While I've done a lot of research on how the two engines work and learned a great deal, there isn't a lot of code at the end of the day. Besides those four commands, what else could I implement between now and the end of March (the deadline for this project)?

Also, if I upload the code here would any of you be able to help me debug and test it? While there are no actual errors, there are a number of things that happen that I can't figure out, like a Playmaker variable being Null but acting as if it were holding the most recent non-null value, and Fungus blocks that stop executing commands once they call a state transition.

Thanks again in advance.

**Comment by chrisgregan** [comment]
Hey Nilihum, sounds like you've made a lot of progress! Happy to take a look at the code. Would you be able to send it in as a pull request on github? To make it easier to review please only commit in changes in the PR that are needed for the new feature (gets hard to figure out what's going on if there's a lot of test code or unrelated changes in there!).

**Comment by chrisgregan** [comment]
...also I have a licensed copy of PlayMaker here so no need to commit that in (as it's not open source).

**Comment by Nilihum** [comment]
I've made a commit consisting of the for main scripts needed. I think, anyway. I don't have much experience with Github.

**Comment by chrisgregan** [comment]
So you committed those changes into your fork of the Fungus repo right? Next step is to send me a pull request so I can review the changes and merge if everything is ok.
https://help.github.com/articles/creating-a-pull-request/

It takes a while to figure out the flow with git / github but once you've done it once or twice it becomes second nature :)

**Comment by Nilihum** [comment]
Yeah, I did that and sent a request. I also clicked "Merge" and now it says "Pull request successfully merged and closed. You’re all set—the Fungus-Playmaker-Integration branch can be safely deleted."

Did that do it automatically or did I make a mistake?

**Comment by Nilihum** [comment]
Just tried again there, did it work?

**Comment by chrisgregan** [comment]
Yep that did it. I'm finished for today though, hopefully get a chance to look through it in the morning!

**Comment by Phenomen** [comment]
@Nilihum I've tested latest Playmaker and your Fungus addon. I can write variables from Playmaker into Fungus flowchart but can't write variables from Fungus into Playmaker global variables (checked your example scene, values become null)

**Comment by Nilihum** [comment]
Thanks for the feedback, @Phenomen. Were you writing  into the Playmaker global variables from the Fungus Variables or the from the Inspector? The former is a known bug that I intend to patch after my exams. Unfortunately, until then, you're either going to have to use the fields in inspector for Write to Playmaker, or use the ReceiveFungusVariable script to have Playmaker read in the Fungus variables to a local Playmaker variable.

**Comment by astraldata** [comment]
Wow. Glad to see this got some love! Thanks guys! 

Wondering did that bug ever got patched, @Nilihum?

**Comment by Nilihum** [comment]
Sorry I haven't been around, @astraldata. The last time I posted I was finishing my final semester in college and was absolutely swamped with assignments, exams, and write up. Ever since then I've been searching to get a job, so this project kind of fell off of my radar. 

However, I just want to say thank you to you, @chrisgregan, and anyone else who gave me support or advice on this. Because of you guys I was able to make this Integration what is is and I will be officially graduating in Novemeber! So thanks to everyone here and at snozbot, I couldn't have done this without you guys.
